### PR TITLE
dyno: add and resolve enums

### PR DESCRIPTION
**Pull request by DanilaFe**
This PR adds enumeration types and params to dyno. This is mainly motivated by the fact that ranges use an enum `BoundedRangeType` as a param, and thus to resolve ranges we need to know how to handle enums. The changes are roughly as follows:

* Add a new `EnumType` subclass to `Type`.
* Also add an `EnumParam` to represent type-level occurrences of enum elements. Where other `*Param` classes refer to the exact value of the number they represent, the `EnumParam` refers to the AST ID that declares the represented element.
    * The conversion to an `Immediate` is left unimplemented at this time, since it seems like such conversions are mostly used for folding primitive operations.
 * Resolve references to enums in code!

In short, the following program now resolves fine:

```Chapel
enum color { red, green, blue }
var x = color.red;
```

### CODE REVIEW

File: compiler/dyno/include/chpl/types/EnumType.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  static const owned<EnumType>& getEnumType(Context* context, ID id, UniqueString name);
+
+ protected:
+  bool contentsMatchInner(const Type* other) const override {
+    auto otherEnum = (const EnumType*) other;
+    return id_ == otherEnum->id_ && name_ == otherEnum->name_;
+  }
+
+  void markUniqueStringsInner(Context* context) const override {
+    name_.mark(context);
```
Comment (mppf) [inline compiler/dyno/include/chpl/types/EnumType.h:45]: Need to also call id_.mark since it contains a UniqueString

File: compiler/dyno/lib/resolution/Resolver.cpp [hunk 2]
```diff
@@ -1890,6 +1890,32 @@ void Resolver::exit(const Dot* dot) {
...
+
+    LookupConfig config = LOOKUP_DECLS | LOOKUP_INNERMOST;
+    auto enumScope = scopeForId(context, receiver.toId());
+    auto vec = lookupNameInScope(context, enumScope,
+                                 /* receiverScope */ nullptr,
+                                 dot->field(), config);
+    ResolvedExpression& r = byPostorder.byAst(dot);
+    if (vec.size() == 0 || vec.size() > 1 || vec[0].numIds() > 1) {
+      // too many or no candidates. The dot expression doesn't specify a valid element.
+      // TODO should be report an error to the user here?
```
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/dyno/lib/resolution/Resolver.cpp:1908]: I think the other case needs to allow for the possibility of function overloading (but it should probably return ErroneousType if there are 0 matches).

**Comment by dlongnecke-cray** [comment]
Michael pointed out some fixups to make. I think this looks great, was a bit confused about why we have getBoundedRangeTypeType (the method itself, not the name). Other than that, looks good to me.

File: compiler/dyno/include/chpl/types/EnumType.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  /** Get an enum type for an ID. */
+  static const EnumType* get(Context* context, ID id, UniqueString name);
+
+  /**
+    Get the type of BoundedRangeType.
+
+    The duplication is intentional. BoundedRangeType is an enum
+    from the standard library. This returns its type.
+   */
+  static const EnumType* getBoundedRangeTypeType(Context* context);
```
[the remark this answers is not in the document]
Comment (DanilaFe) [reply compiler/dyno/include/chpl/types/EnumType.h:62]: This would be _potentially_ useful when manually constructing the `range` type (this PR is a dependency). The code is unused right now, though, and it would probably be best to throw it out for the time being.

[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply compiler/dyno/include/chpl/types/EnumType.h:62]: Yeah I'd be fine with removing this for now and then just constructing the `BoundedRangeType` on the spot where it is needed.
